### ISSUE TITLE
chap03_Linear_Regression_regression_model

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#22]
"""
ver1.x 이용
- X(1) -> Y
- 손실함수(loss function) : 오차 반환 함수
- 모델 최적화 알고리즘: 경사 하강법 알고리즘(GD, Adam)
 -> 모델 학습 : 최적의 기울기, 절편 -> loss 값이 0에 수렴
"""

import tensorflow.compat.v1 as tf # ver1.x
tf.disable_v2_behavior() # ver2.x 사용 안함
import numpy as np
print(tf.__version__) # 2.2.0

### 1. X, Y 변수 정의
x_data = np.array([1,2,3]) # 입력 data
y_data = np.array([2,4,6]) # 출력 data

### 2. X, Y 변수 정의
X = tf.placeholder(dtype=tf.float32, shape = [None]) # x_data공급
Y = tf.placeholder(dtype=tf.float32, shape = [None]) # y_data공급

### 3. a,b 변수 정의 : 
-최초값은 난수로 시작하지만 최적화 객체를 통해서 최적값으로 향해 간다.
a = tf.Variable(tf.random_normal([1])) # 기울기
b = tf.Variable(tf.random_normal([1])) # 절편

### 4. 식 정의
model = tf.multiply(X, a) + b # 예측치 : 회귀방정식

-오차
err = Y - model

-손실 함수
loss = tf.reduce_mean(tf.square(err))

### 5. 최적화 객체 : 기울기, 절편 수정
optimizer = tf.train.GradientDescentOptimizer(0.1) # 학습률 = 0.1
train = optimizer.minimize(loss) # 손실 최소화 : 최적의 기울기, 절편 수정

init = tf.global_variables_initializer() # a,b 위해서 초기화

### 6. 반복학습
with tf.Session() as sess :
    sess.run(init) # 변수 초기화 : a, b
    a_val, b_val = sess.run([a, b])
    print("최초 기울기와 절편")
    print("a = {}, b ={}".format(a_val, b_val)) # a = [0.57575905], b =[2.1578994]
    
    feed_data = {X : x_data, Y : y_data}
    
    -반복학습 : 50회
    for step in range(50) :
        _, loss_val = sess.run([train, loss], feed_dict = feed_data)
        print("step=", (step+1), "loss=", loss_val)
        a_val_up, b_val_up = sess.run([a,b])
        print("수정된a = {}, b ={}".format(a_val_up, b_val_up))
        '''
        step= 1 loss= 32.815918
        수정된a = [2.3210888], b =[-0.06973732]
        
        step= 50 loss= 0.00043357772
        수정된a = [2.0236027], b =[-0.05365467]
        '''
